### PR TITLE
bpo-31884 subprocess set priority on windows

### PR DESCRIPTION
**Pull request by JamesGKent**
adds the required constants for setting process priority to the _winapi module
imports and makes available those constants in the subprocessing module

https://bugs.python.org/issue31884

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by eryksun** [comment]
You could also add the following creation flags: 

    CREATE_NO_WINDOW (allocate a console without a window)
    DETACHED_PROCESS (do not inherit or allocate a console)
    CREATE_DEFAULT_ERROR_MODE
    CREATE_BREAKAWAY_FROM_JOB

The new constants should be documented in section "5.4.1. Constants", with a note about what was added in 3.7.

**Comment by JamesGKent** [comment]
Makes sense, however i thought it was already possible to allocate a console without a window via the startupinfo argument?

**Comment by eryksun** [comment]
When a console application initializes (i.e. the entry point of kernelbase.dll), if it doesn't inherit a console  (i.e. a connection handle for an instance of conhost.exe), it allocates a new console and uses the process `STARTUPINFO` to configure it. This can be used to set the "WindowStation\Desktop"; screen buffer size and default fill attribute; and the window title, size, position, and whether or not to show (hide) the window. Currently we only support a small subset of `STARTUPINFO`, including the `wShowWindow` value.

Use the creation flag `CREATE_NEW_CONSOLE` to force the allocation of a new console instead of inheriting the parent's console. Note that the Secondary Logon service (i.e. `CreateProcessWithLogonW` and `CreateProcessWithTokenW`) always sets this creation flag when it calls `CreateProcessAsUser`. That's why runas.exe creates a new console. The CMD shell's `start` command also uses this flag, unless the `/b` option is used to run a console application in the 'background' (i.e. attached to the current console, but without waiting for it to exit). 

Use `DETACHED_PROCESS` to prevent automatically attaching to a console. The standard handles will initially be `NULL` for a process created with this flag, unless they're redirected to another device. Some console applications assume valid standard handles, so generally you should redirect the standard handles to the NUL character device (i.e. `subprocess.DEVNULL`) when using this flag. A process without an automatically attached console can get one manually by calling `AllocConsole` or `AttachConsole`. 

An issue with using `DETACHED_PROCESS` is that it doesn't prevent grandchild processes from automatically allocating a visible console. To avoid any visible console windows in the default case (i.e. no process in the tree is manually allocating a console), you can either use `CREATE_NEW_CONSOLE`  with a `STARTUPINFO` that hides the console window or use the `CREATE_NO_WINDOW` flag, which makes the process allocate a windowless console. Either way, by default this console will be inherited by child processes. `GetConsoleWindow` returns `NULL` when attached to a console that has no window. In contrast, if it has a hidden window, you can pass the window handle to `ShowWindow` to reveal it. A console without a window or with a hidden window is otherwise functional as far as the API is concerned. 

It's an error to specify both `CREATE_NEW_CONSOLE` and `DETACHED_PROCESS`, and if either is used `CREATE_NO_WINDOW` is ignored.

**Comment by JamesGKent** [comment]
thanks for the thorough explanation, i didn't realize the nuanced combinations of flags that are possible with this API and the effects this has for further child processes, i'll add to the pull request with these new flags and documentation.

**Comment by JamesGKent** [comment]
@eryksun can you clarify where the documentation needs to be ammended?

**Review by vstinner — COMMENTED** [review]
You have to document these constants in the "Constants" section of Doc/library/subprocess.rst.

You have to add ".. versionadded:: 3.7" in the doc of added constants. Maybe add a subsection for Windows priority constants?

**Comment by vstinner** [comment]
test_subprocess fails on Windows (see AppVeyor):

```
  File "C:\projects\cpython\lib\test\test_subprocess.py", line 2900, in test__all__
    self.assertEqual(exported, possible_exports - intentionally_excluded)
AssertionError: Items in the first set but not the second:
'CREATE_BREAKAWAY_FROM_JOB'
'DETACHED_PROCESS'
'CREATE_DEFAULT_ERROR_MODE'
'CREATE_NO_WINDOW'
```

File: Doc/library/subprocess.rst [hunk 1]
```diff
@@ -516,8 +516,24 @@ functions.
...
+   :data:`ABOVE_NORMAL_PRIORITY_CLASS`
+   :data:`BELOW_NORMAL_PRIORITY_CLASS`
+   :data:`HIGH_PRIORITY_CLASS`
+   :data:`IDLE_PRIORITY_CLASS`
+   :data:`NORMAL_PRIORITY_CLASS`
+   :data:`REALTIME_PRIORITY_CLASS`
+   :data:`CREATE_NO_WINDOW`
+   :data:`DETACHED_PROCESS`
+   :data:`CREATE_DEFAULT_ERROR_MODE`
+   :data:`CREATE_BREAKAWAY_FROM_JOB`
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: I'm not sure about this "versioadded" section. I suggest to remove it.

File: Doc/library/subprocess.rst [hunk 2]
```diff
@@ -850,7 +866,64 @@ The :mod:`subprocess` module exposes the following constants.
    on the subprocess.
 
    This flag is ignored if :data:`CREATE_NEW_CONSOLE` is specified.
+   
+.. data:: ABOVE_NORMAL_PRIORITY_CLASS
+
+   A :class:`Popen` ``creationflags`` parameter to specify that a new process
+   will have an above average priority.
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: You have to add ".. versionadded:: 3.7" here: same for each added constant below.

File: Doc/library/subprocess.rst [hunk 3]
```diff
@@ -850,7 +866,64 @@ The :mod:`subprocess` module exposes the following constants.
...
+  A :class:`Popen` ``creationflags`` parameter to specify that a new process
+  does not inherit the error mode of the calling process. Instead, the new
+  process gets the default error mode.
+  This feature is particularly useful for multithreaded shell applications
+  that run with hard errors disabled.
+
+.. data:: CREATE_BREAKAWAY_FROM_JOB
+
+  A :class:`Popen` ``creationflags`` parameter to specify that a new process
+  is not associated with the job.
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: Please add an empty line.

File: Doc/library/subprocess.rst [hunk 4]
```diff
@@ -516,8 +516,24 @@ functions.
 
    If given, *startupinfo* will be a :class:`STARTUPINFO` object, which is
    passed to the underlying ``CreateProcess`` function.
-   *creationflags*, if given, can be :data:`CREATE_NEW_CONSOLE` or
-   :data:`CREATE_NEW_PROCESS_GROUP`. (Windows only)
+   *creationflags*, if given, can be
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: This change seems unrelated and I preferred the previous formatting.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Doc/library/subprocess.rst (hunk 4 above)
Comment (JamesGKent) [reply Doc/library/subprocess.rst]: i'm not sure how this is unrelated? this shows where the added constants should be used?

Comment (vstinner) [reply Doc/library/subprocess.rst]: Ah, when I wrote my comment, the doc was formatted differently.

Ok, now I suggest to format these constants as a list:

```
* a
* b
* c
* ...
```

(correctly indented for the current section)

Try to build the documentation to see how it's rendered in HTML. On Unix, it's "make -C Doc html". I don't know how to build it on Windows :-(

File: Doc/library/subprocess.rst [hunk 5]
```diff
@@ -839,18 +851,87 @@ The :mod:`subprocess` module exposes the following constants.
...
    A :class:`Popen` ``creationflags`` parameter to specify that a new process
    group will be created. This flag is necessary for using :func:`os.kill`
    on the subprocess.
 
    This flag is ignored if :data:`CREATE_NEW_CONSOLE` is specified.
-
+  
+.. data:: ABOVE_NORMAL_PRIORITY_CLASS
+  
+   .. versionadded:: 3.7
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: The versionadded markup is usually added after the text, and surrounded by an empty line (above and after).

File: Doc/library/subprocess.rst [hunk 6]
```diff
@@ -839,18 +851,87 @@ The :mod:`subprocess` module exposes the following constants.
...
 .. data:: CREATE_NEW_CONSOLE
-
+  
    The new process has a new console, instead of inheriting its parent's
    console (the default).
 
 .. data:: CREATE_NEW_PROCESS_GROUP
-
+  
+   .. versionadded:: 3.7
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: This change seems wrong: this constant was already available on Python 3.6.

File: Doc/library/subprocess.rst (hunk 4 above)
Comment (JamesGKent) [reply Doc/library/subprocess.rst]: unfortunately i don't currently have access to a unix box (windows 10 broke my dual boot) and currently on a work pc so don't even have git client, doing this through the web interface... so i can't build the docs to see how they look

File: Doc/library/subprocess.rst [hunk 7]
```diff
@@ -843,6 +855,8 @@ The :mod:`subprocess` module exposes the following constants.
    The new process has a new console, instead of inheriting its parent's
    console (the default).
 
+   .. versionadded:: 3.7
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: That's wrong: the flag already existed in Python 3.6.

File: Doc/library/subprocess.rst [hunk 8]
```diff
@@ -851,6 +865,86 @@ The :mod:`subprocess` module exposes the following constants.
 
    This flag is ignored if :data:`CREATE_NEW_CONSOLE` is specified.
 
+   .. versionadded:: 3.7
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: That's wrong: the flag already existed in Python 3.6.

File: Doc/library/subprocess.rst [hunk 9]
```diff
@@ -516,8 +516,20 @@ functions.
...
+      * :data:`BELOW_NORMAL_PRIORITY_CLASS`
+      * :data:`HIGH_PRIORITY_CLASS`
+      * :data:`IDLE_PRIORITY_CLASS`
+      * :data:`NORMAL_PRIORITY_CLASS`
+      * :data:`REALTIME_PRIORITY_CLASS`
+      * :data:`CREATE_NO_WINDOW`
+      * :data:`DETACHED_PROCESS`
+      * :data:`CREATE_DEFAULT_ERROR_MODE`
+      * :data:`CREATE_BREAKAWAY_FROM_JOB`
+
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: A single empty line is enough at the end of this list.

File: Doc/library/subprocess.rst [hunk 10]
```diff
@@ -843,6 +855,8 @@ The :mod:`subprocess` module exposes the following constants.
    The new process has a new console, instead of inheriting its parent's
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: The title of the current section is " Constants". I checked: *all* constants documented here as specific to Windows. I propose to rename the section to "Windows Constants". Can you please do that?

File: Doc/library/subprocess.rst (hunk 4 above)
Comment (vstinner) [reply Doc/library/subprocess.rst]: >  *creationflags*, if given, can be

I checked in MSDN: in fact, you can specify *multiple* flags at once:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms686331(v=vs.85).aspx

dwFlags: "A bitfield that determines whether certain STARTUPINFO members are used when the process creates a window. **This member can be one or more of the following values.**"

---

By the way, please add a column after "can be" ("can be:"), and an empty line after.

I propose: "can be one or more of the following flags:".

**Review by vstinner — COMMENTED** [review]
Thanks. The overall change looks good to me, but I still have a few comments on the documentation which can be enhanced.

**Comment by vstinner** [comment]
The doc job of Travis CI failed with:

```
[1] library/subprocess.rst:532: trailing whitespace
[1] library/subprocess.rst:915: trailing whitespace
[1] library/subprocess.rst:922: trailing whitespace
```

Please remove also these trailing spaces ;-)

**Review by vstinner — APPROVED** [review]
LGTM. Thank you for all updates! And I'm sorry for being so much pedantic about this PR ;-)

**Comment by vstinner** [comment]
@zooba: Nice, subprocess now will get the constants that you used in https://github.com/haypo/perf to set the priority ;-)

**Comment by JamesGKent** [comment]
i don't mind the being pedantic, its my first time contributing to a project this structured, but i wouldn't have it any other way. there is a reason that pythons docs are comprehensive and useful...
i only realized this was missing due to a question on stackoverflow and thinking "i'm sure you should be able to do that" defined the constants in the python script and it just worked, so figured they ought to be added to the module

**Comment by vstinner** [comment]
PR merged, thank again ;-) I changed the commit title.
